### PR TITLE
Add relay capability.

### PR DESCRIPTION
**Pull request by Daniel-Oberlin**
I added a feature to your program to support "relay" functionality.  I need to be able to wake my machine remotely over the internet, but my router (like many) will not pass through a broadcast packet to my LAN.  The relay capability will listen for a regular WOL packet addressed to the machine, and then resend it as broadcast to the LAN.  I'm running this on a Rasbperry Pi on my LAN, so that I can wake other machines remotely.

I tried to add this feature with minimal intrusion to the original code.

### CODE REVIEW

File: cli/index.js [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
+    relay = require('../').relay,
     isMACValid = require('../').isMACValid;
 
 program
     .usage('<mac>')
     .option('--ip [ip]', 'IP Address [255.255.255.255] of target computer', '255.255.255.255')
     .option('--host [host]', 'Hostname of the target computer')
     .option('--port [port]', 'Port [9]', 9)
+    .option('--relay', 'Listen for packet and then broadcast it')
+    .option('--relayport [port]', 'Listen for relay packet on this port')
```
Comment (sbmaxx) [inline cli/index.js]: You should validate user input here. For example, it should be set with value (not boolean), maybe casted to number. I didn't write whose checks initially, but i think this should be done — https://github.com/tj/commander.js#coercion :)

File: cli/index.js [hunk 2]
```diff
@@ -1,13 +1,16 @@
 var program = require('commander'),
     chalk = require('chalk'),
     wake = require('../'),
+    relay = require('../').relay,
```
Comment (sbmaxx) [inline cli/index.js]: we don't need two requires here. why not just `require('wake').relay`? also i think, that we could change api to use it like `wake({ relay })`

File: cli/index.js [hunk 3]
```diff
@@ -22,8 +25,19 @@ if (!isMACValid(mac)) {
...
-});
+if (program.relay) {
+    relay(mac, {
+        ip: program.ip,
+        host: program.host,
+        port: program.port,
+        relayport: program.relayport
+    });
+} else {
+    wake(mac, {
```
Comment (sbmaxx) [inline cli/index.js]: as i mention before, we can use the same `wake` method and pass `relay` and `relayport` optiosn

File: index.js [hunk 4]
```diff
@@ -102,3 +100,50 @@ module.exports.isMACValid = function(mac) {
 
     return !(mac.length != 2 * MAC_BYTES || mac.match(/[^a-fA-F0-9]/));
 }
+
+module.exports.relay = function(mac, params) {
+
+        getIP(params).then(function(ip) {
```
Comment (sbmaxx) [inline index.js]: we can extend `wake` method. for example:

```js
module.exports = function wake(...) {
    getIP(params).then(function(ip) {
         if (params.relay) {
             wakeThroughRelay();
         } else {
             wake();
         }
    });
};
```

File: index.js [hunk 5]
```diff
@@ -102,3 +100,50 @@ module.exports.isMACValid = function(mac) {
...
+            
+            var magicPacket = createMagicPacket(mac);
+            listensocket = dgram.createSocket('udp4');
+            
+            listensocket.on('error', (err) => {
```
Comment (sbmaxx) [inline index.js]: if you are using es6, please change node version in package.json

File: index.js [hunk 6]
```diff
@@ -102,3 +100,50 @@ module.exports.isMACValid = function(mac) {
...
+              console.log(`server error:\n${err.stack}`);
```
Comment (sbmaxx) [inline index.js]: `console.error` :)

File: index.js [hunk 7]
```diff
@@ -102,3 +100,50 @@ module.exports.isMACValid = function(mac) {
...
+              listensocket.close();
+            });
+            
+            listensocket.on('message', (msg, rinfo) => {
+            
+              if (msg.equals(magicPacket))
```
Comment (sbmaxx) [inline index.js]: i don't have any codestyle linters, but i prefer this style
```js
if () {
}
```

File: index.js [hunk 8]
```diff
@@ -102,3 +100,50 @@ module.exports.isMACValid = function(mac) {
...
+              {
+                var broadcastsocket = dgram.createSocket(net.isIPv6(ip) ? 'udp6' : 'udp4');
+                
+                broadcastsocket.once('listening', function() {
```
Comment (sbmaxx) [inline index.js]: why you don't use arrow functions here? :)

File: index.js [hunk 9]
```diff
@@ -102,3 +100,50 @@ module.exports.isMACValid = function(mac) {
...
+                });
+        
+                console.log('Broadcasting magic packet to %s.', chalk.blue(mac));
+
+                broadcastsocket.send(magicPacket, 0, magicPacket.length, params.port, BROADCAST, function(err) {
+                    if (err) {
+                        console.log(chalk.red('Sorry ;('));
+                        console.error(err);
+                    } else {
+                        console.log('%s. Your computer is awakening right now...', chalk.green('All\'s fine'));
```
Comment (sbmaxx) [inline index.js]: I think we should reuse strings in both wake & relay methods

File: index.js [hunk 10]
```diff
@@ -102,3 +100,50 @@ module.exports.isMACValid = function(mac) {
...
+                });
+              }
+            });
+            
+            listensocket.on('listening', () => {
+              const address = listensocket.address();
+              console.log(`server listening ${address.address}:${address.port}`);
+            });
+            
+            listensocket.bind(params.relayport);
```
Comment (sbmaxx) [inline index.js]: maybe just `socket` is enough for var name?

**Review by sbmaxx — COMMENTED** [review]
Daniel, thanks for you contribution! 

I wrote some comments. Feel free to ask if you don't understand smth. Also, i can help with refactoring, if you need my help:)

**Comment by Daniel-Oberlin** [comment]
Hi Roman,

Thanks for the feedback and comments!

This is the first code I've ever written for Node, so I appreciate the time
that you've taken.  I'll make some updates and get back to you, but could
be a week or so.

Best regards,
Dan

**Comment by sbmaxx** [comment]
Please don't hesitate to contact me if you have any question or need help :)

File: cli/index.js (hunk 2 above)
Comment (Daniel-Oberlin) [reply cli/index.js]: Ok

File: cli/index.js (hunk 1 above)
Comment (Daniel-Oberlin) [reply cli/index.js]: I looked at the documentation, and found this example which is similar:

...
  .option('-v, --verbose', 'A value that can be increased', increaseVerbosity, 0)
...

Is this what you had in mind?  I can add that to the code.

File: cli/index.js (hunk 3 above)
Comment (Daniel-Oberlin) [reply cli/index.js]: I can refactor to a single method.  I thought that the wake vs. relay were significantly different that they merited distinct methods, but I'm happy to change that.

File: index.js (hunk 4 above)
Comment (Daniel-Oberlin) [reply index.js]: Ok

File: index.js (hunk 5 above)
Comment (Daniel-Oberlin) [reply index.js]: I got this code from an example that I found.

I'm quite new to JS, and not too familiar with what has been updated in ES6 and how that affects the node version.  Could you tell me what about this is requires ES6, and the implications to node version?

File: index.js (hunk 6 above)
Comment (Daniel-Oberlin) [reply index.js]: Yep!  :-)

File: index.js (hunk 7 above)
Comment (Daniel-Oberlin) [reply index.js]: Sure, I'll change to be consistent with the rest of the code.

File: index.js (hunk 8 above)
Comment (Daniel-Oberlin) [reply index.js]: Just looked them up and read about them!  I'll make the change.

File: index.js (hunk 10 above)
Comment (Daniel-Oberlin) [reply index.js]: Ok.

File: index.js (hunk 9 above)
Comment (Daniel-Oberlin) [reply index.js]: Yep, will do that in the refactor.

File: cli/index.js (hunk 1 above)
Comment (Daniel-Oberlin) [reply cli/index.js]: Specifically, for the --relay option

File: index.js (hunk 5 above)
Comment (Daniel-Oberlin) [reply index.js]: I see, you mean the arrow function here.  Do you have any objection to updating the node version, then?

File: index.js (hunk 10 above)
Comment (Daniel-Oberlin) [reply index.js]: We have two sockets, so I wanted to be very explicit and name them listensocket and broadcastsocket.  That ok?

**Comment by Daniel-Oberlin** [comment]
Thanks,

Just one question.  This forced me to learn about arrow functions, and I
see now that they are a feature of es6.

You wrote this comment:

"if you are using es6, please change node version in package.json"

But I'm not sure exactly what would need to change.  I did a little reading
and couldn't figure it out.  Can you give me a quick pointer here?  I have
v8.7.0 of node installed locally, but I don't understand what would need to
change in package.json or why...?

Thanks,
Dan

File: index.js (hunk 5 above)
Comment (sbmaxx) [reply index.js]: I think, we can easily upgrade to latest node.js version :)

File: index.js (hunk 10 above)
Comment (sbmaxx) [reply index.js]: Hm, sounds reasonable

**Comment by sbmaxx** [comment]
About node.js version — https://docs.npmjs.com/files/package.json#engines

**Comment by sbmaxx** [comment]
> This is the first code I've ever written for Node

I'm just curious,  what is your main programming language?

**Comment by Daniel-Oberlin** [comment]
Thanks for the info!

I've done a lot of programming in various languages, most of them C-based - C++, Java, C#...

I'm excited to be getting into node for some personal projects.

**Comment by Daniel-Oberlin** [comment]
Hi Roman,

I've refactored and made all of the changes in my branch.  I'm not sure what to do next, would I create another pull request, or do I somehow indicate that I am finished and then you would use this existing pull request?

Thanks,
Dan

**Comment by sbmaxx** [comment]
I'll take a deep look today (maybe tomorrow). This PR is fine.

**Comment by sbmaxx** [comment]
LGTM

**Comment by sbmaxx** [comment]
@Daniel-Oberlin thx for your contribution. Landed in https://www.npmjs.com/package/wake-on-lan `wake-on-lan@0.1.0`

**Comment by Daniel-Oberlin** [comment]
You're welcome!

**Pull request merged**